[geometry] add support for polyhedral surface and TIN

ptitjano opened the pull request:
## Description

This PR adds support for the `POLYHEDRALSURFACE` and `TIN` simple features by introducing `QgsPolyhedralSurface`  and `QgsTriangulatedSurface`: 
- `QgsPolyhedralSurface` inherits from `QgsSurface` 
- `QgsTriangulatedSurface` inherits from `QgsPolyhedralSurface` as a special case which only contains triangles.  

This approach is similar to the one used by GDAL: https://gdal.org/en/latest/development/rfc/rfc64_triangle_polyhedralsurface_tin.html

Everything should work as expected: create a new layer or edit an existing one, WKT/WKB conversion, etc..

Some previous workarounds are also removed. For example:
- TIN is now the multipart of a triangle
- there is no need to convert a polyhedral surface or a TIN to a multi polygon when importing PostGIS data

Funded by CEA/DAM, CP4SC, Oslandia

ptitjano committed:
qgis: Add support for polyhedralsurface type

ptitjano committed:
geometry: Add polyhedralsurface support

ptitjano committed:
qgsgeometryfactory: Add support for qgspolyhedralsurface

ptitjano committed:
qgis: Add support for TIN type

ptitjano committed:
geometry: Add triangulatedsurface support

ptitjano committed:
qgsgeometryfactory: Add support for qgstriangulatedsurface

ptitjano committed:
qgsgeometry: Use QgsPolyhedralSurface in fromBox3D

ptitjano committed:
qgssymbol: Add support for qgspolyhedralsurface

ptitjano committed:
qgssymbol: Add support for qgstriangulatedsurface

ptitjano committed:
qgspolygon3dsymbol_p: Add support for polyhedralsurface

This allows to process a QgsPolyhedralSurface geometry to display it.

ptitjano committed:
qgsgeometry: Add polygon to polyhedralsurface conversion in coerceToType

This allows to edit a vector layer of polyhedral surface to add new
polygon features.

ptitjano committed:
qgsgeos: Add support for polyhedralsurface in asGeos

This allows to convert a `QgsPolyhedralSurface` to a `MultiPolygon`
geos geometry.

ptitjano committed:
qgsgeos: Add support for TIN in asGeos

This allows to convert a `QgsTriangulatedSurface` to a `MultiPolygon`
geos geometry.

lbartoletti committed:
PostgreSQL: Add new geometry types

Allow PolyhedralSurface and TIN in PostgreSQL table.
Add a new GeometryColumnType and Capabilities : Surfaces.

lbartoletti committed:
PostGIS/ogr: Remove conversion of PolyhedralSurface/TIN to MultiPolygon

QGIS now natively supports PostGIS PolyhedralSurface and TIN
geometries. This commit removes the workaround that converted
them to MultiPolygon during loading.

lbartoletti committed:
qgsnewmemorylayerdialog: Add support for qgspolyhedralsurface

lbartoletti committed:
qgsnewmemorylayerdialog: Add support for TIN

lbartoletti committed:
Triangle: allow to create and edit triangle geometry type

ptitjano committed:
qgssurface: Do not define surfaceToPolygon

`surfaceToPolygon` is only useful for `QgsCurvePolygon` which inherits
from `QgsSurface`. The next commit will introduce
`QgsPolyhedralSurface` which also inherits from `QgsSurface`. For
`QgsPolyhedralSurface`, it does not make sense to have
`surfaceToPolygon`.

By moving `surfaceToPolygon` definition in `QgsCurvePolygon` this
allows to keep the functionality.

lbartoletti commented:
This is the last part to fully support Triangle and welcome to PolyhedralSurface/TIN. Here, we support all SF geometries.

WKT/WKB tested against SFCGAL (which only one to fully support these geometries), used in PostGIS and GDAL.

Fantastic job done by @ptitjano !

rouault commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -1772,6 +1779,37 @@ QVector<QgsGeometry> QgsGeometry::coerceToType( const Qgis::WkbType type, double
+    const QgsGeometry source = newGeom;
+    for ( auto part = source.const_parts_begin(); part != source.const_parts_end(); ++part )
+    {
+      std::unique_ptr< QgsAbstractGeometry > partAbstractGeometry( ( *part )->clone() );
```
I may miss something, but the clone of the part doesn't seem necessary, since you don't actually add it to another component, but cast it as a polygon, and clone that polygone.

ptitjano replied:
You're right. Fixed.

rouault commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -1772,6 +1779,37 @@ QVector<QgsGeometry> QgsGeometry::coerceToType( const Qgis::WkbType type, double
+  {
+    std::unique_ptr< QgsTriangle > triangle = std::make_unique< QgsTriangle >();
+    const QgsGeometry source = newGeom;
+    if ( QgsPolygon *polygon = qgsgeometry_cast< QgsPolygon * >( newGeom.constGet()->clone() ) )
```
there must be a memory leak here. polygon is not freed. What should be done instead is to *not* clone polygon, but clone polygon->exteriorRing() since setExteriorRing() takes ownership of it

ptitjano replied:
Indeed. Fixed.

rouault commented on src/core/geometry/qgsgeometrycollection.h:
```diff
@@ -334,7 +334,13 @@ class CORE_EXPORT QgsGeometryCollection: public QgsAbstractGeometry
+        return nullptr;
+
+      const Qgis::WkbType flatType = QgsWkbTypes::flatType( geom->wkbType() );
+      if ( QgsWkbTypes::isMultiType( geom->wkbType() )
```
I would argue that isMultiType() should not return true for PolyhedralSurface and TIN if they can't be cast to GeometryCollection. At least that is what GDAL does

ptitjano replied:
While working on it, I was not sure on the correct approach. Since, we're trying to have an approach similar to GDAL. This should be changed. Thanks for the suggestion.

nyalldawson replied:
Yeah, there's a TON of places where we check if a geometry isMultiType and then assume it can be safely static_cast as a collection. Let's not break that assumption here.

ptitjano replied:
In `WkbEntries`, one need to define the `multiType` of `polyhedraSurface` and `TIN`. What would be the appropriate multi type: a `QgsMultiPolygon`, a `QgsGeometryCollection`? 

SFCGAL uses a `Solid`, `MultiSolid` logic but we do not have this logic in QGIS.

nyalldawson replied:
I'd say multipolygon

ptitjano replied:
`QgsPolyhedralSurface` and `QgsTriangulatedSurface` are now single types. I have also removed the logic to consider `TIN` as the multi type of `QgsTriangle`.

rouault commented on src/core/geometry/qgsgeometryeditutils.cpp:
```diff
@@ -105,6 +107,20 @@ Qgis::GeometryOperationResult QgsGeometryEditUtils::addPart( QgsAbstractGeometry
+      {
+        triangulatedSurface->addPatch( triangle );
+      }
+    }
```
missing return Qgis::GeometryOperationResult::Success in the success code path and Qgis::GeometryOperationResult::InvalidInputGeometryType if the part is not a triangle ?

ptitjano replied:
Indeed. Fixed.

rouault commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+{
+  if ( &p != this )
+  {
+    clearCache();
```
```suggestion
    clear();
```

ptitjano replied:
This is mostly a copy-paste from `QgsCurvePolygon`: https://github.com/qgis/QGIS/blob/2cda36e46a6d67937ec3dbd94a3116e71493c78f/src/core/geometry/qgscurvepolygon.cpp#L87

Does that mean that the code in `QgsCurvePolygon` is wrong?

rouault replied:
> Does that mean that the code in `QgsCurvePolygon` is wrong?

likely. This should be a full clear, not just a clear cache

ptitjano replied:
It turns out that it is not needed. See: https://github.com/qgis/QGIS/pull/58587

rouault commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+
+QgsAbstractGeometry *QgsPolyhedralSurface::childGeometry( int index ) const
+{
+  return mPatches.at( index - 1 );
```
Likely a copy&paste left-over from QgsCurvePolygon::childGeometry()  (shows also a test might be needed for that)
```suggestion
  return mPatches.at( index );
```

ptitjano replied:
This probably comes from a copy-paste! I have also added a unit test.

rouault commented on src/core/geometry/qgspolyhedralsurface.h:
```diff
@@ -0,0 +1,341 @@
+
+      const Qgis::WkbType flatType = QgsWkbTypes::flatType( geom->wkbType() );
+      if ( flatType == Qgis::WkbType::PolyhedralSurface
+           || flatType == Qgis::WkbType::TIN
+           || flatType == Qgis::WkbType::Polygon
+           || flatType == Qgis::WkbType::Triangle )
```
you can't cast a Polygon* or Triangle* to a PolyhedralSurface*
```suggestion
           || flatType == Qgis::WkbType::TIN )
```

rouault commented on src/core/geometry/qgstriangulatedsurface.cpp:
```diff
@@ -0,0 +1,343 @@
+{
+  if ( &p != this )
+  {
+    clearCache();
```
```suggestion
    clear();
```

rouault commented on src/core/geometry/qgstriangulatedsurface.h:
```diff
@@ -0,0 +1,214 @@
+        return nullptr;
+
+      const Qgis::WkbType flatType = QgsWkbTypes::flatType( geom->wkbType() );
+      if ( flatType == Qgis::WkbType::TIN
+           || flatType == Qgis::WkbType::Triangle )
```
```suggestion
      if ( flatType == Qgis::WkbType::TIN )
```

ptitjano replied:
Fixed.

rouault commented on src/core/providers/qgsabstractdatabaseproviderconnection.h:
```diff
@@ -533,6 +533,7 @@ class CORE_EXPORT QgsAbstractDatabaseProviderConnection : public QgsAbstractProv
       SinglePoint = 1 << 5,            //!< Supports single point types (as distinct from multi point types) \since QGIS 3.28
       SingleLineString = 1 << 6,       //!< Supports single linestring types (as distinct from multi line types) \since QGIS 3.28
       SinglePolygon = 1 << 7,          //!< Supports single polygon types (as distinct from multi polygon types) \since QGIS 3.28
+      Surfaces = 1 << 8,              //!< Supports surface (PolyhedralSurface, TIN) types (as distinct from multi polygon types) \since QGIS 3.40
```
Surfaces also encompasses Polygon or CurvePolygon, so may be bit more specific with:
```suggestion
      PolyhedralSurfaces = 1 << 8,              //!< Supports surface (PolyhedralSurface, TIN) types (as distinct from multi polygon types) \since QGIS 3.40
```

ptitjano replied:
Done.

ptitjano committed:
qgsoracleprovider: Ignore polyhedralsurface type

This is not handled at the moment.

ptitjano committed:
qgsoracleprovider: Ignore TIN type

This is not handled at the moment.

nyalldawson commented on src/core/qgis.h:
```diff
@@ -213,6 +213,8 @@ class CORE_EXPORT Qgis
       CurvePolygon = 10, //!< CurvePolygon
       MultiCurve = 11, //!< MultiCurve
       MultiSurface = 12, //!< MultiSurface
+      PolyhedralSurface = 15, //!< PolyhedralSurface
```
```suggestion
      PolyhedralSurface = 15, //!< PolyhedralSurface \since QGIS 3.40
```

elsewhere too

ptitjano replied:
Done.

nyalldawson commented on src/core/providers/qgsabstractdatabaseproviderconnection.h:
```diff
@@ -533,6 +533,7 @@ class CORE_EXPORT QgsAbstractDatabaseProviderConnection : public QgsAbstractProv
       SinglePoint = 1 << 5,            //!< Supports single point types (as distinct from multi point types) \since QGIS 3.28
       SingleLineString = 1 << 6,       //!< Supports single linestring types (as distinct from multi line types) \since QGIS 3.28
       SinglePolygon = 1 << 7,          //!< Supports single polygon types (as distinct from multi polygon types) \since QGIS 3.28
+      PolyhedralSurfaces = 1 << 8,     //!< Supports surface (PolyhedralSurface, TIN) types (as distinct from multi polygon types) \since QGIS 3.40
```
```suggestion
      PolyhedralSurfaces = 1 << 8,     //!< Supports polyhedral surfaces (PolyhedralSurface, TIN) types (as distinct from multi polygon types) \since QGIS 3.40
```

ptitjano replied:
Fixed.

nyalldawson commented on src/core/geometry/qgswkbtypes.h:
```diff
@@ -204,18 +208,17 @@ class CORE_EXPORT QgsWkbTypes
-        // until we support TIN types, use multipolygon
         case Qgis::WkbType::Triangle:
-          return Qgis::WkbType::MultiPolygon;
+          return Qgis::WkbType::TIN;
```
This is confusing/inconsistent. If we are considering tin/polyhedral types as single types (which I agree with), then we shouldn't be returning them as the multi type here.

ptitjano replied:
Removed.

nyalldawson commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -259,7 +263,7 @@ QgsGeometry QgsGeometry::fromBox3D( const QgsBox3D &box )
   }
 
-  std::unique_ptr< QgsMultiPolygon > multiPolygon = std::make_unique< QgsMultiPolygon >();
+  std::unique_ptr< QgsPolyhedralSurface > polyhedralSurface = std::make_unique< QgsPolyhedralSurface >();
```
I think this behaviour should be noted in the docs for fromBox3D -- eg "\returns a 2D polygon geometry if the box is purely 2d, otherwise returns a polyhedral surface geometry."

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -1772,6 +1778,37 @@ QVector<QgsGeometry> QgsGeometry::coerceToType( const Qgis::WkbType type, double
+    const QgsGeometry source = newGeom;
+    for ( auto part = source.const_parts_begin(); part != source.const_parts_end(); ++part )
+    {
+      std::unique_ptr< const QgsAbstractGeometry > partAbstractGeometry( *part );
+      if ( QgsPolygon *polygon = qgsgeometry_cast< QgsPolygon * >( partAbstractGeometry.get() ) )
+      {
+        polySurface->addPatch( polygon->clone() );
+      }
```
```suggestion
      if ( const QgsPolygon *polygon = qgsgeometry_cast< const QgsPolygon * >( *part ) )
      {
        polySurface->addPatch( polygon->clone() );
      }
```

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -1849,6 +1886,17 @@ bool QgsGeometry::convertToMultiType()
   }
 
+  // TriangulatedSurface does not inherit from QgsGeometryCollection
+  if ( QgsWkbTypes::flatType( d->geometry->wkbType() ) == Qgis::WkbType::Triangle )
```
Again, we've got an inconsistency here -- calling convertToMultiType on a triangle will return a geometry which is NOT a multitype! :scream:

ptitjano replied:
Removed.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.h:
```diff
@@ -0,0 +1,340 @@
+/**
+ * \ingroup core
+ * \class QgsPolyhedralSurface
+ * \brief polyhedral surface geometry type
```
Can you add more detail here? There should be sufficient information included so that it's immediately clear what a polyhedral surface type is and how/why it differs from regular polygon types.

(Also make sure the docs are properly capitalised)

ptitjano replied:
Fixed.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+  mPatches.reserve( multiPolygon->numGeometries() );
+  for ( int i = 0; i < multiPolygon->numGeometries(); ++i )
+  {
+    mPatches.append( multiPolygon->polygonN( i ) );
```
This is certainly going to crash -- you've marked this method as transferring ownership of the multiPolygon, but at the end of this method the multiPolygon still owns all the parts returned by polygonN... which are now also owned by this object.

I would instead remove the ownership transferral from this method, make the multiPolygon argument const, and use a clone here instead.

ptitjano replied:
Fixed.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.h:
```diff
@@ -0,0 +1,340 @@
+      wkt = wkt.left( 1000 ) + QStringLiteral( "..." );
+    QString str = QStringLiteral( "<QgsPolyhedralSurface: %1>" ).arg( wkt );
+    sipRes = PyUnicode_FromString( str.toUtf8().constData() );
+    % End
```
It'd be great to implement __len__ here, and ideally __getitem__ too. That'd make it possible to directly iterate over a QgsPolyhedralSurface in python

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+
+{
+  mWkbType = p.mWkbType;
+
```
```suggestion
  mPatches.reserve( p.mPatches.size() );
```

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+  if ( &p != this )
+  {
+    QgsSurface::operator=( p );
+
```
```suggestion
    mPatches.reserve( p.mPatches.size() );
```

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+
+json QgsPolyhedralSurface::asJsonObject( int ) const
+{
+  QgsDebugError( QStringLiteral( "geojson format does not support PolyhedralSurface geometry" ) );
```
Shouldn't we treat this as a GeoJSON multipolygon then?

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+  multiLine->reserve( mPatches.size() );
+  for ( QgsPolygon *polygon : mPatches )
+  {
+    QgsAbstractGeometry *polygonBoundary = polygon->boundary();
```
unique_ptr

ptitjano replied:
fixed.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+  for ( QgsPolygon *patch : mPatches )
+  {
+    // exterior ring
+    QgsCurve *exteriorRing = static_cast< QgsCurve *>( patch->exteriorRing()->snappedToGrid( hSpacing, vSpacing, dSpacing, mSpacing, removeRedundantPoints ) );
```
unique_ptr

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+      return nullptr;
+    }
+
+    QgsPolygon *gridifiedPatch = new QgsPolygon;
```
unique_ptr

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+      if ( !interiorRing )
+        continue;
+
+      QgsCurve *gridifiedInterior = static_cast< QgsCurve * >( interiorRing->snappedToGrid( hSpacing, vSpacing, dSpacing, mSpacing, removeRedundantPoints ) );
```
unique_ptr

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+  std::unique_ptr< QgsPolyhedralSurface > simplifiedGeom = std::make_unique< QgsPolyhedralSurface >();
+  for ( QgsPolygon *polygon : mPatches )
+  {
+    simplifiedGeom->addPatch( polygon->simplifyByDistance( tolerance )->surfaceToPolygon() );
```
leaks the geometry returned by simplifyByDistance

ptitjano replied:
Fixed.

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+
+QgsMultiSurface *QgsPolyhedralSurface::toCurveType() const
+{
+  QgsMultiSurface *multiSurface = new QgsMultiSurface();
```
unique_ptr

nyalldawson commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -0,0 +1,997 @@
+
+QgsMultiPolygon *QgsPolyhedralSurface::toMultiPolygon() const
+{
+  QgsMultiPolygon *multiPolygon = new QgsMultiPolygon();
```
unique_ptr

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgstriangulatedsurface.h:
```diff
@@ -0,0 +1,213 @@
+/**
+ * \ingroup core
+ * \class QgsTriangulatedSurface
+ * \brief triangulated surface geometry type
```
Same here, can you expand these docs to give a lot more detail about this geometry type?

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgstriangulatedsurface.cpp:
```diff
@@ -0,0 +1,343 @@
+
+json QgsTriangulatedSurface::asJsonObject( int ) const
+{
+  QgsDebugError( QStringLiteral( "geojson format does not support TIN geometry" ) );
```
Similar here -- i'd just return as a multipolygon

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgstriangulatedsurface.cpp:
```diff
@@ -0,0 +1,343 @@
+      return nullptr;
+    }
+
+    QgsTriangle *gridifiedTriangle = new QgsTriangle;
```
unique_ptr

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgstriangulatedsurface.cpp:
```diff
@@ -0,0 +1,343 @@
+    if ( !triangle )
+      continue;
+
+    QgsCurve *exteriorRing = qgsgeometry_cast< QgsCurve *>( triangle->exteriorRing()->snappedToGrid( hSpacing, vSpacing, dSpacing, mSpacing, removeRedundantPoints ) );
```
unique_ptr

ptitjano replied:
Done.

nyalldawson commented:
Great work @ptitjano  -- thanks for closing this large gap in our geometry coverage!

benoitdm-oslandia reviewed (approved):
LGTM!

ptitjano committed:
qgswkbtypes: Remove comment regarding triangle multiType

`TIN` support has been added but it is a single type. Therefore, the
multi type of a Triangle is a MultiPolygon.
